lib/slots.ts

```diff
@@ -39,8 +42,52 @@ export const NANOSECONDS = 'slot-nanoseconds';
 // Calendar
 export const CALENDAR_ID = 'slot-calendar-identifier';
 
+interface Slots {
+  // Instant
+  [EPOCHNANOSECONDS]: bigInt.BigInteger; // number? JSBI?
+
+  // TimeZone
+  [TIMEZONE_ID]: string;
+
+  // DateTime, Date, Time, YearMonth, MonthDay
+  [ISO_YEAR]: number;
+  [ISO_MONTH]: number;
+  [ISO_DAY]: number;
+  [ISO_HOUR]: number;
+  [ISO_MINUTE]: number;
+  [ISO_SECOND]: number;
+  [ISO_MILLISECOND]: number;
+  [ISO_MICROSECOND]: number;
+  [ISO_NANOSECOND]: number;
+  [CALENDAR]: Temporal.Calendar;
```

Comment:
I actually think this is CalendarProtocol, because non-built-in calendars can also be stored.

Reply:
This doesn't work, as it triggers ~50 errors about assignability: 

`Argument of type 'CalendarProtocol' is not assignable to parameter of type 'Calendar'.ts(2345)`

I've started to see this show up elsewhere when I turn on more of the compiler options. It makes sense - `Calendar` is intentionally different from `CalendarProtocol` from the perspective of users, and it seems everywhere we accept a `CalendarProtocol` we do call `ToTemporalCalendar` which should convert a `CalendarProtocol` to a `Calendar`. At least in my branch, that method isn't typed and in most of the places we do this conversion it looks a bit something like

```
constructor(..., calendar: Temporal.CalendarProtocol | string = ES.GetISO8601Calendar()) {
   ...
    calendar = ES.ToTemporalCalendar(calendar);

    // Note: if the arguments are not passed,
    //       ToIntegerThrowOnInfinity(undefined) will have returned 0, which will
    //       be rejected by RejectISODate in CreateTemporalDateSlots. This check
    //       exists only to improve the error message.
    if (arguments.length < 3) {
      throw new RangeError('missing argument: isoYear, isoMonth and isoDay are required');
    }

    ES.CreateTemporalDateSlots(this, isoYear, isoMonth, isoDay, calendar);
  }
```

which doesn't compile as `CreateTemporalDateSlots` shouldn't accept `CalendarProtocol`, only `Calendar`. If this is rewritten to no longer reassign the parameter, and ToTemporalCalendar is typed correctly, then I think I can say with certainty that all user calendars (and likely, all user TimeZones) are converted from the `Protocol` version to the internal version.

I was wondering if whether some sort of similar difference in types would be useful to enforce that these coercion methods are actually called and the results used.

Reply:
Yeah, this is the param reassignment problem. Once we fix that, then I think most of these errors should go away-- any others can probably be addressed with type assertions or (worst case) additional variable renaming.

Too bad that TS doesn't narrow types from reassignment. Here's me making the case for why it should: https://github.com/microsoft/TypeScript/issues/45870#issuecomment-932649448

Reply:
ToTemporalCalendar can't narrow the type, because of plain object calendars. At first glance I think it is actually correct that CreateTemporalDateSlots should accept CalendarProtocol, for that reason.